### PR TITLE
Inject GCP IAM Credentials into GCE PD CSI Driver Test Namespace by Default

### LINKED ISSUE #62561: GCE PD CSI Driver E2E test should automatically inject driver GCP credentials
Currently there is no way for the GCE P driver to automatically get GCP credentials in the E2E test framework so that it has the permissions to Create/Delete/Attach/Detach volumes on GCP.

The workaround to run the test currently is to manually create a service account for the project, then expose it as a secret in the namespace the test is running in.

Might be able to solve with TokenRequest API: https://github.com/kubernetes/features/issues/542
Or maybe there is a way to automatically inject the credentials in CI that I am not aware of.

/kind enhancement
/sig storage

### PR DESCRIPTION
**Pull request by davidz627**
**What this PR does / why we need it**:
This PR automatically injects GCP IAM Credentials into the test namespace for the GCE PD CSI Driver to use when running. This allows the test to run completely autonomously.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #62561

/assign @saad-ali @msau42 

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: which block is CI and which one is desktop?

Also in CI, does the test runner run in the same project where the cluster is created?

Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: you're right, I'm going to need to hack on this some more - the project here is the one where the test runner is run, not necessarily the one where the cluster is created.

File: test/e2e/storage/csi_objects.go
Comment (BenTheElder) [inline test/e2e/storage/csi_objects.go]: In the CI this is generating keys which creates a lot of warning emails for us, and is leaving them behind, we at the very least probably need to extend the tooling to purge these keys after testing 😬

Comment (BenTheElder) [reply test/e2e/storage/csi_objects.go]: Been thinking about this more (and getting more warnings), is it possible to identify the agent better? Right now I don't think we can differentiate between this tooling generating a key and someone trying to exfil a key via a naughty "test" or similar 🤔 I feel like end users might have similar concerns about these warnings.

Comment (BenTheElder) [reply test/e2e/storage/csi_objects.go]: OK, we can at least tell that it is indeed the PR's service account creating this account, but it's unfortunate that we can't trivially tie it back to a job. maybe a very explicit log message so we can trivially track it back to the tests's logs?

Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: alright so i've updated it to shred the downloaded file and delete the key. Not sure what you mean by very explicit log message but I have this one:
`framework.Logf("Creating service account key for iam account %v in project %v at location %v", serviceAccount, projectID, saFile)`
let me know if you have something else in mind

**Comment by davidz627** [comment]
@msau42 ready for review

File: test/e2e/storage/csi_objects.go
Comment (BenTheElder) [reply test/e2e/storage/csi_objects.go]: Can we use a pre-existing key in a secret or not run this in the default e2es? I'm concerned that the infra maintainers will have to either:
- filter out key download notifications (= bad we should be watching for these...)
- actually sift through a deluge of service account key download emails

We're getting an alert from GCP every time this runs, delayed by about one day.

Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: There's no way to "get" a service account key, you can only get the actual private key data on creation of the key so thats necessary.

A compromise here may be to have it still gated behind a `[Feature: _]` thing so that it doesn't run in CI by default, and turn it on when external pod identity goes beta/GA and we can use that.

@saad-ali, @msau42  do you have any opinions on how important this is to run in CI by default?

Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: @msau42 suggested maybe pre-creating and downloading one service key for each project and saving them somewhere "safe" to be used by this test. @BenTheElder is there a known/existing mechanism for doing something like this?

Comment (BenTheElder) [reply test/e2e/storage/csi_objects.go]: we just discussed this offline, this is what I meant :-) we have keys that we use for CI resources, PR for anyone following along is https://github.com/kubernetes/test-infra/pull/9320

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: Do we really need to clean up the namespace at the beginning of the test?  Every test case should run in a new namespace, so it should already be clean

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: when do you need multiple service accounts?

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: why beta?

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: do we still need to have this auto generation logic?  If you're running it manually, we could also require you setup the key separately using the existing scripts.

File: test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml
Comment (msau42) [inline test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml]: don't we still need to specify a service account?

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: Recommended deployment of GCE PD uses two service accounts, one for node, the other for controller

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: not necessary, but it doesn't hurt.

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: I can change this back to v1. There was a bug at some point I thought was due to podsecuritypolicy being only supported in beta

File: test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml
Comment (davidz627) [reply test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml]: it's always going to be specified by a generated service account. Removing this dummy default value causes it to fail faster if not specified

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: hmm yes, we could but then the test wouldn't run properly automatically. It is definitely a tradeoff. I prefer it to be fully automated like this but I could also be convinced the other way

Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: consider the case where someone just wants to run all storage tests, its weird to just have the CSI GCE ones fail because the user wasn't present when prompted to do some manual actions.

Comment (msau42) [reply test/e2e/storage/csi_objects.go]: Can we just require that the env variable must be set to run the gce pd tests?  It could be a bit alarming to have the test unexpectently create and download keys from the project.  That means the user also has to run the e2e tests with a gcp account that has a lot of permissions

File: test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml
Comment (msau42) [reply test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml]: You mean every pod automatically gets a service account injected into it?

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: good point, i'll have it spit out a message to set the environment variable

**Comment by davidz627** [comment]
@msau42 well its way simpler now, PTAL

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: is this going to work if multiple tests run in parallel?

File: test/e2e/storage/csi_objects.go
Comment (msau42) [inline test/e2e/storage/csi_objects.go]: do we need to copy it? are we modifying it?

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: added a UUID-ed directory

File: test/e2e/storage/csi_objects.go
Comment (davidz627) [reply test/e2e/storage/csi_objects.go]: no, but it needs to be called `cloud-sa.json` and this is the cleanest way of doing that. Another method could be to change the testing manifest to use whatever the file in `E2E_GOOGLE_APPLICATION_CREDENTIALS` but thats arguably more hacky as it requires digging into the manifest, looping through its containers, looping through the right containers env vars and changing the value of the `GOOGLE_APPLICATION_CREDENTIALS` variable.

**Comment by davidz627** [comment]
@msau42 comments addressed. thanks for your thorough review, this PR is 110% better with your help :)

**Comment by davidz627** [comment]
/hold cancel

@msau42 could you do final review

**Review by saad-ali — APPROVED** [review]
/approve

**Comment by saad-ali** [comment]
Let's cherry pick to 1.12 too

**Pull request merged**
